satellite/gc/service.go

```diff
@@ -0,0 +1,171 @@
+// Copyright (C) 2019 Storj Labs, Inc.
+// See LICENSE for copying information.
+
+package gc
+
+import (
+	"context"
+	"sync"
+	"time"
+
+	"github.com/zeebo/errs"
+	"go.uber.org/zap"
+	monkit "gopkg.in/spacemonkeygo/monkit.v2"
+
+	"storj.io/storj/internal/sync2"
+	"storj.io/storj/pkg/bloomfilter"
+	"storj.io/storj/pkg/overlay"
+	"storj.io/storj/pkg/pb"
+	"storj.io/storj/pkg/storj"
+	"storj.io/storj/pkg/transport"
+	"storj.io/storj/satellite/metainfo"
+	"storj.io/storj/uplink/piecestore"
+)
+
+var (
+	// Error defines the gc service errors class
+	Error = errs.Class("gc service error")
+	mon   = monkit.Package()
+)
+
+// Config contains configurable values for garbage collection
+type Config struct {
+	Interval time.Duration `help:"the time between each send of garbage collection filters to storage nodes" releaseDefault:"168h" devDefault:"10m"`
+	Enabled  bool          `help:"set if garbage collection is enabled or not" releaseDefault:"true" devDefault:"true"`
+	// value for InitialPieces currently based on average pieces per node
+	InitialPieces     int     `help:"the initial number of pieces expected for a storage node to have, used for creating a filter" releaseDefault:"400000" devDefault:"10"`
+	FalsePositiveRate float64 `help:"the false positive rate used for creating a filter" releaseDefault:"0.1" devDefault:"0.1"`
+	ConcurrentSends   int     `help:"the number of nodes to concurrently send bloom filters to" releaseDefault:"1" devDefault:"1"`
+}
+
+// Service implements the garbage collection service
+type Service struct {
+	log          *zap.Logger
+	Loop         *sync2.Cycle
+	metainfoloop *metainfo.Loop
+	transport    transport.Client
+	overlay      overlay.DB
+	config       Config
+
+	lastPieceCounts map[storj.NodeID]int
+
+	pieceCountsMutex sync.Mutex
+	pieceCounts      map[storj.NodeID]int
+}
+
+// RetainInfo contains info needed for a storage node to retain important data and delete garbage data
+type RetainInfo struct {
+	Filter       *bloomfilter.Filter
+	CreationDate time.Time
+	count        int
+}
+
+// NewService creates a new instance of the gc service
+func NewService(log *zap.Logger, transport transport.Client, overlay overlay.DB, loop *metainfo.Loop, config Config) *Service {
+	// TODO retrieve piece counts from overlay (when there is a column for them)
+	// var lastPieceCounts atomic.Value
+	// lastPieceCounts.Store(map[storj.NodeID]int{})
+	lastPieceCounts := make(map[storj.NodeID]int)
+
+	return &Service{
+		log:              log,
+		Loop:             sync2.NewCycle(config.Interval),
+		metainfoloop:     loop,
+		transport:        transport,
+		overlay:          overlay,
+		lastPieceCounts:  lastPieceCounts,
+		config:           config,
+		pieceCountsMutex: sync.Mutex{},
```

Comment:
Wouldn't `pieceCountsMutex` be initialized even without this statement?

Reply:
mutex unnecessary now. Also, I think you are right